// angular-auth-oidc-client/esm2022/lib/auto-login/auto-login-all-routes.guard.mjs
import { Injectable, inject } from '@angular/core';
import { Router, } from '@angular/router';
import { map, switchMap, take } from 'rxjs/operators';
import { CheckAuthService } from '../auth-state/check-auth.service';
import { ConfigurationService } from '../config/config.service';
import { LoginService } from '../login/login.service';
import { AutoLoginService } from './auto-login.service';
import * as i0 from "@angular/core";
/**
 * @deprecated Please do not use the `AutoLoginAllRoutesGuard` anymore as it is not recommended anymore, deprecated and will be removed in future versions of this library. More information [Why is AutoLoginAllRoutesGuard not recommended?](https://github.com/damienbod/angular-auth-oidc-client/issues/1549)
 */
export class AutoLoginAllRoutesGuard {
    constructor() {
        this.autoLoginService = inject(AutoLoginService);
        this.checkAuthService = inject(CheckAuthService);
        this.loginService = inject(LoginService);
        this.configurationService = inject(ConfigurationService);
        this.router = inject(Router);
    }
    canLoad() {
        const url = this.router
            .getCurrentNavigation()
            ?.extractedUrl.toString()
            .substring(1) ?? '';
        return checkAuth(url, this.configurationService, this.checkAuthService, this.autoLoginService, this.loginService);
    }
    canActivate(route, state) {
        return checkAuth(state.url, this.configurationService, this.checkAuthService, this.autoLoginService, this.loginService);
    }
    canActivateChild(route, state) {
        return checkAuth(state.url, this.configurationService, this.checkAuthService, this.autoLoginService, this.loginService);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.8", ngImport: i0, type: AutoLoginAllRoutesGuard, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.0.8", ngImport: i0, type: AutoLoginAllRoutesGuard, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.8", ngImport: i0, type: AutoLoginAllRoutesGuard, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
function checkAuth(url, configurationService, checkAuthService, autoLoginService, loginService) {
    return configurationService.getOpenIDConfiguration().pipe(switchMap((config) => {
        const allConfigs = configurationService.getAllConfigurations();
        return checkAuthService.checkAuth(config, allConfigs).pipe(take(1), map(({ isAuthenticated }) => {
            if (isAuthenticated) {
                autoLoginService.checkSavedRedirectRouteAndNavigate(config);
            }
            if (!isAuthenticated) {
                autoLoginService.saveRedirectRoute(config, url);
                loginService.login(config);
            }
            return isAuthenticated;
        }));
    }));
}
